BootloaderCommonPkg/Library/IppCryptoPerfLib/IppCryptoPerfLib.c

```diff
@@ -0,0 +1,133 @@
+
+/* @file
+*
+*  Copyright (c) 2024, Intel Corporation. All rights reserved.<BR>
+*
+*  SPDX-License-Identifier: BSD-2-Clause-Patent
+*
+*/
+
+#include <Library/CryptoLib.h>
+#include <Library/ShellLib.h>
+#include <Library/TimeStampLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/SecureBootLib.h>
+#include <Base.h>
+#include <Library/BootloaderCommonLib.h>
+#include <Library/LoaderPerformanceLib.h>
+
+#define SIZE_10MB (SIZE_8MB + SIZE_2MB)
+#define MAX_ITER 4
+#define MAX_DIGEST_BUFFER_SIZE SHA384_DIGEST_SIZE
+#define CRYPTO_MEASURE_POINT_ID 0x5000
+struct IterationLength {
+    UINT32 Length;
+    UINT8 Desc[7];
+};
+
+const struct IterationLength Iter_premem[] = {{ SIZE_1KB,   "1 KB"   },
+                                       { SIZE_32KB,  "32 KB"  },
+                                       { SIZE_64KB,  "64 KB"  },
+                                       { SIZE_128KB, "128 KB" }};
+
+const struct IterationLength Iter_postmem[] = {{ SIZE_64KB,  "64 KB"  },
+                                       { SIZE_256KB, "256 KB" },
+                                       { SIZE_1MB,   "1 MB"   },
+                                       { SIZE_10MB,  "10 MB"  }};
+
+enum Algorithm {
+    SHA256,
+    SHA384,
+    MAX_ALGO
+};
+
+enum BootStage {
+    PREMEM,
+    POSTMEM,
+    MAX_STAGE
+};
+
+static const char* AlgorithmName (enum Algorithm Alg) {
+    if (Alg == SHA256)
+        return "SHA256";
+    else if (Alg == SHA384)
+        return "SHA384";
+    else
+        return NULL;
+}
+
+VOID EFIAPI CryptoPerfTest (
+    IN UINT8 Stage
+)
+{
+    UINT16 MeasurePointId;
+    UINT8 DigestBuffer[MAX_DIGEST_BUFFER_SIZE];
+    UINT8 *Data;
+    UINT8 Index;
+    enum Algorith Alg;
+    const struct IterationLength *Iter;
+
+    if (Stage == PREMEM) {
+        Iter = Iter_premem;
+        Data = (UINT8*)(UINTN)FixedPcdGet32(PcdStage1ABase);
+    } else if (Stage == POSTMEM) {
+        Iter = Iter_postmem;
+        Data = (UINT8*) AllocatePool (Iter[MAX_ITER - 1].Length);
+        GenerateRandomNumbers(Data, Iter[MAX_ITER - 1].Length);
+    } else {
+        return;
+    }
+
+    for (Alg = 0; Alg < MAX_ALGO; Alg++) {
+        MeasurePointId = (UINT16)(CRYPTO_MEASURE_POINT_ID + (Stage << 8) + (Alg << 4));
+        for (Index = 0; Index < MAX_ITER; Index++) {
+            if (Alg == SHA256) {
+                    AddMeasurePoint (MeasurePointId);
+                    Sha256 (Data, Iter[Index].Length, DigestBuffer);
+                    AddMeasurePoint (MeasurePointId+1);
+            } else if (Alg == SHA384) {
+                    AddMeasurePoint (MeasurePointId);
+                    Sha384 (Data, Iter[Index].Length, DigestBuffer);
+                    AddMeasurePoint (MeasurePointId+1);
+            }
+            MeasurePointId += 2;
+        }
+    }
+    if (Stage == POSTMEM && Data != NULL)
+        FreePool(Data);
+}
+
+VOID EFIAPI CryptoPerfTestPrintResult (
```

Comment:
This function is not required since perf command could print all the measured data.

Reply:
Since every delta is in micro seconds and not in milliseconds I am printing it here.

Reply:
got it.